### ISSUE TITLE
Writing & Reading persons `vehicles` attributes from XML

### ISSUE THREAD

**Issue by Nitnelav** [matsim-org/matsim-libs#1916]
I'm using the emissions contrib and to do so I generated a fleet of vehicles for my synethic population.

Then, I'm using `QSimConfigGroup.VehiclesSource.fromVehiclesData` and I set the `vehiclesFile` parameter.
It's quite unclear (and undocumented?), but then vehicles must be assigned to persons for every mode.
That means writing something like that : 

```java
      if (config.qsim().getVehiclesSource() == QSimConfigGroup.VehiclesSource.fromVehiclesData) {
          for (Person person : scenario.getPopulation().getPersons().values()) {
              Id<Vehicle> vehicleId = Id.createVehicleId(person.getId());
              Map<String, Id<Vehicle>> modeVehicle = new HashMap<>();
              modeVehicle.put("car", vehicleId);
              VehicleUtils.insertVehicleIdsIntoAttributes(person, modeVehicle);
          }
      }
```

So far so good.

 - Problem : 
But now I want to assign the vehicles to persons directly as I generate my vehicle fleet, in python (using https://github.com/eqasim-org/ile-de-france) outside of matsim java code. 
That means adding the `vehicles` attribute to persons directly in the XML file. But the  `vehicles` attribute is actually a `Map<String, Id<Vehicle>>` and there is no [AttributeConverter](https://github.com/matsim-org/matsim-libs/blob/master/matsim/src/main/java/org/matsim/utils/objectattributes/AttributeConverter.java) for this.

 - Solution A :
There is a `Map<String, String>` [Converter](https://github.com/matsim-org/matsim-libs/blob/master/matsim/src/main/java/org/matsim/utils/objectattributes/attributeconverters/StringStringMapConverter.java) that should be straightforward to use for vehicles.
That would give us : 
```xml
	<person id="1030955">
		<attributes>
			<attribute name="age" class="java.lang.Integer">61</attribute>
			<attribute name="sex" class="java.lang.String">m</attribute>
			<attribute name="vehicles" class="java.util.Collections$UnmodifiableMap">{"car":"1030955"}</attribute>
		</attributes>
		<plan selected="yes">
			<activity type="home" link="180085" facility="home_458069" x="352759.54" y="6688605.21" >
			</activity>
		</plan>
	</person>
``` 

 - Solution B : 
The "vehicles that belong to person X" seems to be a core element of an agent definition, it would make sense to add it in its own WML tags like : 
```xml
	<person id="1030955">
		<attributes>
			<attribute name="age" class="java.lang.Integer">61</attribute>
			<attribute name="sex" class="java.lang.String">m</attribute>
		</attributes>
                <vehicles>
                        <vehicle mode="car" id="1030955" />
                </vehicles>
		<plan selected="yes">
			<activity type="home" link="180085" facility="home_458069" x="352759.54" y="6688605.21" >
			</activity>
		</plan>
	</person>
``` 

What do you think ?

**Comment by sebhoerl** [comment]
Another thought that just came to my mind. Could we have a class `org.matsim.*.PersonVehicles` and then have an `AttributeConverter` for that?

**Comment by Nitnelav** [comment]
Yes that works.
I did manage to get this result using a `PersonVehicles` Class and a `PersonVehiclesAttributeConverter` Class.

```java
	<person id="1030955">
		<attributes>
			<attribute name="age" class="java.lang.Integer">61</attribute>
			<attribute name="sex" class="java.lang.String">m</attribute>
			<attribute name="personVehicles" class="org.eqasim.core.misc.PersonVehicles">{"car":"1030955"}</attribute>
		</attributes>
		<plan selected="yes">
			<activity type="home" link="180085" facility="home_458069" x="352759.54" y="6688605.21" >
			</activity>
		</plan>
	</person>			
```

I think I like Solution B better, but it requires larger changes, most importantly a `population_v7.dtd` and its `PopulationWriterHandlerImplV7`.

whatever community thinks is best, I can try to implement it and submit a PR.

**Comment by Janekdererste** [comment]
Although I agree, that vehicles are at the core of our simulation, we tend to put more and more things into attributes instead of finxing it in the schema. This way we are more flexible, when things need to change.

If you put the vehicle information into the person's attribtues, I would think that using the Mechanism for ```Map<String, String>``` should be enough. The attribute value in the above comment looks exactly like when it would have been generated with the ```StringStringMapConverter```. 

To make this work automatically, the ```PrepareForSimImpl.getVehicleTypesForAllNetworkAndMainModes``` would have to be extended anyway. This part could also untangle the ```Map<String, String>``` into a list of vehicle ids for the person.

**Comment by Nitnelav** [comment]
Yes, the `PersonVehiclesAttributeConverter` I wrote actually calls `StringStringMapConverter` internally hence the similarity :) 

Ok for using Person attributes instead of updating the schema. 

Do you prefer using `StringStringMapConverter` only or should we add a couple Classes like `PersonVehicles` and  `PersonVehiclesAttributeConverter` as @sebhoerl suggested ?

**Comment by kainagel** [comment]
Thanks for looking into this.  The `fromVehicles` was at some point introduced because it is most consistent with the matsim design, but it was indeed never programmed out.

I would go with Janek that we would prefer putting it into the attributes.  This is also what we have done at many other places.  Maybe we move it into a typed attribute at some later format change, but in the end it seems best to keep formats stable as long as possible.

One additional thing: In places where we had the same problem, we consolidated corresponding static methods into a `*Utils` class.  Look, e.g., into the `*Hbefa*` getters and setters in `VehicleUtils`.  Using it in that way also means that it would be easy to inline it later if it ever becomes a typed attribute.  (One part of this is to _not_ expose the attributes keyword, so that people are forced to use the utils method.)

**Comment by Janekdererste** [comment]
Having the type safety would indeed be nice. On the other hand, when using a ```Map<String, String>``` everything is already there. 

The three methods following [VehicleUtils:123](https://github.com/matsim-org/matsim-libs/blob/772f7d112669586fb45e58660fde3a85e910bbd6/matsim/src/main/java/org/matsim/vehicles/VehicleUtils.java#L123) will do what you need already.

**Comment by Nitnelav** [comment]
I wrote something that works using a `PersonVehicles` class, you can check it out here : https://github.com/matsim-org/matsim-libs/pull/1922

**Comment by kainagel** [comment]
The additional attribute converters need to be registered with the readers/writers.  If you are programming a contrib, this is a bit tricky.  That is the main reason why from a contrib we use generalized attribute converters.

Since these changes concern the core, the additional attribute converters can be registered permanently, and in consequence the typesafe approach is nicer.

I would prefer to have the additional static methods in `<containerName>Utils'.  I.e. in this case `PopulationUtils`.  Makes them easier to find in the long run.
